### DISCUSSION TITLE
How to properly close a background child process in a shutdown event?

### DISCUSSION THREAD

**Discussion by clifflau1120** [fastapi/fastapi#7285] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from multiprocessing import Process, Value

from fastapi import FastAPI

app = FastAPI()

terminate_flag = Value('h', 0)


def work_hard(terminate_flag: Value):
    while terminate_flag.value == 0:
        # CPU-bound calculations
        pass


@app.get('/hello-world')
def hello_world():
    return {
        'message': 'hello world'
    }


process = Process(target=work_hard, args=(terminate_flag,))


@app.on_event('startup')
def startup_event():
    process.start()


@app.on_event('shutdown')
def shutdown_event():
    terminate_flag.value = 1
    process.join()

```

### Description

I have some CPU-bound calculations known as `work_hard`, which contains a flagged loop to control its flow.

The function is designed to be executed in a child process that starts within FastAPI `startup` event, and joins within FastAPI `shutdown` event.

However, it seems that the FastAPI `shutdown` event does not get executed while `uvicorn` is shutting down by the user pressing Ctrl + C.
It stuck at the log "`Finished server process`" and never really terminates.

I did try to look up for the standard practice of using `multiprocessing` module in FastAPI, but there are no concrete examples about it.
[Concurrency and async/await](https://fastapi.tiangolo.com/async/#concurrency-parallelism-web-machine-learning) section mentioned that "To see how to achieve this parallelism in production see the section about Deployment.", yet [Deployment](https://fastapi.tiangolo.com/deployment/) section has no such examples as well.

Thank you in advance!

### Environment

* OS: Ubuntu 18.04.4 TLS
* FastAPI Version : 0.61.1
* Python version: 3.6.8

### Additional context

```bash
cliff@cliff-pc:~/repo/fastapi-multiprocessing$ uvicorn app.main:app
INFO:     Started server process [24032]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
^CINFO:     Shutting down
INFO:     Finished server process [24032]
```

**Comment by ArcLightSlavik** [comment]
Looks **very** similar to this: https://stackoverflow.com/questions/58133694/graceful-shutdown-of-uvicorn-starlette-app-with-websockets.
I've tried both methods and neither seem to work, maybe you will have better luck:rocket:

**Comment by clifflau1120** [comment]
How do you guys handle parallelism in your FastAPI application by the way?
Using `multiprocessing` is not a must for me, I just wanted an elegant, maintainable way to implement it.

**Comment by ArcLightSlavik** [comment]
I believe the way you do it is by having asynchronous code within FastAPI, and then you deploy with `gunicorn` workers which provide parallelism. 
I am not sure about running `multiprocessing` within FastAPI itself.

**Comment by raphaelauv** [comment]
You can scale using `gunicorn` and set a number of `workers`

Or 

Wrap your code in a docker container running a unique ASGI web server ( `uvicorn` | `hypercorn` ...) process 
and let your docker scheduler (`docker-swarm`| `kubernetes` ... ) scale the number of containers|pods to run.

**Comment by raphaelauv** [comment]
Also to respond to your question :

` FastAPI shutdown event does not get executed while uvicorn is shutting down by the user pressing Ctrl + C.`

No it does 

```python

from fastapi import FastAPI
import uvicorn

async def on_shutdown():
    print("HERE")

if __name__ == '__main__': 
    web_app = FastAPI(docs_url="/", on_shutdown=[on_shutdown])
    uvicorn.run(web_app, host="0.0.0.0", port=8000)
```

give 

```shell
INFO:     Started server process [10579]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
^CINFO:     Shutting down
INFO:     Waiting for application shutdown.
HERE
INFO:     Application shutdown complete.
INFO:     Finished server process [10579]
```

Probably you run the code inside a docker and do not use [tini](https://github.com/krallin/tini)

**Comment by raphaelauv** [comment]
@clifflau1120 You can apparently close this issue. Thank you

**Comment by e0lithic** [comment]
@raphaelauv . I have created a reproducible example where the same issue persists. This is a mini app which creates a central server where your webcam or other IP cameras can be visualized. The requirements are `imutils, imagezmq, fastapi, opencv-python`. I apologize for not being able to create a more standalone example due to lack of time. However, I think this should help in highlighting the issues raised in the ongoing thread. In order to reproduce kindly run the following : 
```bash
# Start the server program
uvicorn fastapp:app --reload
# Visit the URL so that the server initializes itself for the client process.

# On a different terminal run the following : 
# Run the client to send images to the server
python camera_client_0.py
# The mini app should be up and running and your webcam would be live streaming in the browser url.
# Try Ctrl + C ( twice)
# The terminal would get stuck at Finished server proces
```
SourceCode : [exp.tar.gz](https://github.com/tiangolo/fastapi/files/5317007/exp.tar.gz)

Thanks

**Comment by raphaelauv** [comment]
@clifflau1120 

Could you close this issue , thank you.

**Comment by 3tilley** [comment]
Hello, I have the terminal issue as well. I don't know whether this issue should be reopened or a new one created, let me know either way.

I can replicate this issue, and it's pretty annoying! The shutdown hook appears to fire intermittently, and only in the workers, but not in the parent process. I'm not sure whether it should fire in the parent, but it certainly shouldn't block the terminal. It only seems to occur for multiple workers though. See the code snippet below (adapted from the above example). Tweak workers from 1 to 2 to see it not working.

I think this only occurs on Windows, though I've tried cmd and git bash with the same effect (screenshot from git bash):

```
import sys

from fastapi import FastAPI
import uvicorn

async def on_shutdown():
    print("HERE")

app = FastAPI(docs_url="/", on_shutdown=[on_shutdown])

if __name__ == '__main__': 
    workers = int(sys.argv[1])
    print(f"Running with {workers} worker(s)")
    uvicorn.run("main2:app", host="0.0.0.0", port=8001, workers=workers)
```

![image](https://user-images.githubusercontent.com/1138504/116951291-59a8f080-ac7f-11eb-93f4-09bcd5df5206.png)

**Comment by clifflau1120** [comment]
> 
> 
> Hello, I have the terminal issue as well. I don't know whether this issue should be reopened or a new one created, let me know either way.
> 
> I can replicate this issue, and it's pretty annoying! The shutdown hook appears to fire intermittently, and only in the workers, but not in the parent process. I'm not sure whether it should fire in the parent, but it certainly shouldn't block the terminal. It only seems to occur for multiple workers though. See the code snippet below (adapted from the above example). Tweak workers from 1 to 2 to see it not working.
> 
> I think this only occurs on Windows, though I've tried cmd and git bash with the same effect (screenshot from git bash):
> 
> ```
> import sys
> 
> from fastapi import FastAPI
> import uvicorn
> 
> async def on_shutdown():
>     print("HERE")
> 
> app = FastAPI(docs_url="/", on_shutdown=[on_shutdown])
> 
> if __name__ == '__main__': 
>     workers = int(sys.argv[1])
>     print(f"Running with {workers} worker(s)")
>     uvicorn.run("main2:app", host="0.0.0.0", port=8001, workers=workers)
> ```
> 
> ![image](https://user-images.githubusercontent.com/1138504/116951291-59a8f080-ac7f-11eb-93f4-09bcd5df5206.png)

It seems that there is no simple solution to this.
At the end, I broke the workers into standalone programs and communicate with them via sockets.
Certainly not the best way doing it, but it solves my problem.

**Comment by 3tilley** [comment]
@raphaelauv now we've got a repro, are you happy to reopen the issue? If you have any ideas where the issue might be I'm can do a tinker myself

**Comment by raphaelauv** [comment, marked as answer]
"happy"  ..........

just search ! https://github.com/encode/uvicorn/pull/853

```python
import uvicorn
from starlette.applications import Starlette

async def on_shutdown():
    print("HERE")

app = Starlette(debug=True, on_shutdown=[on_shutdown])

if __name__ == "__main__":
    uvicorn.run("toto:app", host="127.0.0.1", port=5000, log_level="debug", workers=2)
````
It's not fastapi, its uvicorn and windows and issue open by clifflau1120 was never about windows.

I will no more reply.

**Comment by tiangolo** [comment]
Thanks for the help and discussion here everyone and thanks @clifflau1120 for coming back to close the issue. ☕ 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
